### PR TITLE
Implement aws.ecs.* resource attributes

### PR DESCRIPTION
**Pull request by mmanciop**
Implemented retrieving the [`aws.ecs.*` resource attributes](https://opentelemetry.io/docs/reference/specification/resource/semantic_conventions/cloud_provider/aws/ecs/) in `go.opentelemetry.io/detectors/aws/ecs` based on the ECS Metadata v4 endpoint.

### CODE REVIEW

File: detectors/aws/ecs/ecs_test.go
Comment (hanyuancheung) [inline detectors/aws/ecs/ecs_test.go:48]:
```suggestion
// succesfully returns resource when process is running on Amazon ECS environment 
// with no Metadata v4.
```

File: detectors/aws/ecs/ecs_test.go
Comment (hanyuancheung) [inline detectors/aws/ecs/ecs_test.go:71]:
```suggestion
// succesfully returns resource when process is running on Amazon ECS environment 
// with Metadata v4.
```

**Comment by mmanciop** [comment]
There's discussion on https://github.com/open-telemetry/opentelemetry-java/pull/4574#issuecomment-1213233736 whether the ARNs like the one this PR builds are actually correct. Until we are use, let's not merge :D

**Comment by mmanciop** [comment]
**Update:** The uncertainty about ARNs seems sorted out

**Comment by bryan-aguilar** [comment]
Can we add test cases that utilize the [fargate examples responses](https://docs.aws.amazon.com/AmazonECS/latest/userguide/task-metadata-endpoint-v4-fargate.html) also?

**Comment by bryan-aguilar** [comment]
Looks like there may be slight differences between fargate and ec2 metadata response. As a sanity check can we add those example responses linked in the comment above to the unit tests?

**Comment by mmanciop** [comment]
> Looks like there may be slight differences between fargate and ec2 metadata response. As a sanity check can we add those example responses linked in the comment above to the unit tests?

Done in 2ef2026553d481e14f3c5c051dc9df5afc8a2bb7. There are no factual differences that this patch cares about.

**Review by bryan-aguilar — APPROVED** [review]
From the ADOT team perspective this looks good to go. 
Any long term concerns with using [AWS ECS Metadata Go](https://github.com/brunoscheufler/aws-ecs-metadata-go) as it describes itself being built for fargate services?

File: detectors/aws/ecs/ecs.go
Comment (Aneurysm9) [inline detectors/aws/ecs/ecs.go:139]: This is [an enumeration](https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/resource/cloud_provider/aws/ecs.yaml#L18-L27) that does not allow custom values.  Can you use one of the [defined values](https://github.com/open-telemetry/opentelemetry-go/blob/main/semconv/v1.12.0/resource.go#L224-L229) for this key?

File: detectors/aws/ecs/ecs.go
Comment (Aneurysm9) [inline detectors/aws/ecs/ecs.go:167]:
```suggestion
		r := regexp.MustCompile(`^arn:aws:ecs:([^:]+):`)
```

File: detectors/aws/ecs/ecs.go
Comment (Aneurysm9) [inline detectors/aws/ecs/ecs.go:171]:
```suggestion
	r := regexp.MustCompile(`^arn:aws:ecs:[^:]+:([^:]+):`)
```

File: detectors/aws/ecs/ecs_test.go
Comment (Aneurysm9) [inline detectors/aws/ecs/ecs_test.go:210]: What is the purpose of this test? It makes no assertions.

Comment (mmanciop) [reply detectors/aws/ecs/ecs_test.go]: Yeah it was spurious, removed with 3490115

File: detectors/aws/ecs/ecs.go
Comment (mmanciop) [reply detectors/aws/ecs/ecs.go:135]: Done in 8e6bf31d

**Comment by Aneurysm9** [comment]
All of the log group/stream fields are optional.  There is no requirement to include all of them if any of them are included.

File: detectors/aws/ecs/ecs.go
Comment (Aneurysm9) [reply detectors/aws/ecs/ecs.go:135]: This has not been changed, only the tests.  This does not ensure that a non-standard value cannot be emitted.

**Comment by mmanciop** [comment]
> All of the log group/stream fields are optional. There is no requirement to include all of them if any of them are included.

IMO, this would wreak havoc with less sophisticated tracing backends that do not have processing capabilities to normalize the attributes, as the same Log stream or group could be represented with different values.

File: detectors/aws/ecs/ecs.go
Comment (mmanciop) [reply detectors/aws/ecs/ecs.go]: Ah, now I understood what you mean. Done in 49156c98.

File: detectors/aws/ecs/ecs.go
Comment (mmanciop) [reply detectors/aws/ecs/ecs.go]: String slices introduced in 4e857dd171637b361c147a96b35f5401f2d3703f. I am not sold on skipping attributes though, see https://github.com/open-telemetry/opentelemetry-go-contrib/pull/2626#issuecomment-1219113163.

**Review by Aneurysm9 — APPROVED** [review]

**Comment by mmanciop** [comment]
Rebased on main, solved a bunch of merge conflicts; @Aneurysm9 care to have a look?

**Comment by Aneurysm9** [comment]
@mmanciop This requires an additional approval from @open-telemetry/go-approvers before it can be merged.

File: detectors/aws/ecs/ecs.go
Comment (dmathieu) [inline detectors/aws/ecs/ecs.go:26]: This repository has 3 stars and 14 commits. I wonder how reliable it can be in the future.
It's also fairly simple. So we can also extract its content and bring it back into this package later on if necessary.

**Review by dmathieu — APPROVED** [review]

**Comment by mmanciop** [comment]
@Aneurysm9 does this ring a bell?

> panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: socket: The requested service provider could not be loaded or initialized. [recovered]
> 	panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: socket: The requested service provider could not be loaded or initialized.

I copied the test code from the AksDetector and the tests pass on my Mac; am I perhaps missing something in the CI configs?

**Pull request merged**
